### PR TITLE
Support for OpenEXR to build additional apps

### PR DESCRIPTION
**Pull request by remia**
This PR is a follow up of a work started by @michdolan a while back. It mainly deals with the OpenImageIO / OpenColorIO circular dependency issue for apps (ocioconvert, ociolutimage and ociodisplay) by allowing the use of OpenEXR to build those when OIIO is not found.

Here is a quick overview of the content:
* Replace `oiiohelpers` by `imageioapphelpers`, this library expose the class `ImageIO` which uses the private implementation idiom to support either OpenEXR or OIIO backend depending on build config,
* Remove support for Half (OpenEXR 2) from the library,
* Uses and requires OpenEXR and Imath latest versions (3.1.5),
* FindOpenEXR.cmake script to allow in source build, currently doesn't support looking for the library by other mean than the `find_package(... CONFIG)` mode as the minimal requested version install a config script,
* Fix a const issue in `CPUProcessor::apply`

There is still some more testing to be done but I prefer to open this early in case there are any decision made above that needs more discussion or revision.

### CODE REVIEW

**Review by doug-walker — APPROVED** [review]
This adds some welcome flexibility, thanks Remi!

File: share/cmake/modules/FindOpenEXR.cmake
Comment (remia) [reply share/cmake/modules/FindOpenEXR.cmake:29]: Good point Doug, I wonder what we should do in case OIIO is not found and we don't find OpenEXR on the system. Currently it will use ExternalProject as mentioned to bring it to the build, unless `OCIO_INSTALL_EXT_PACKAGES` is set to `NONE`. Should we change that to only build it if set to `ALL`? Alternatively use another CMake `option` to explicitly ask for `OpenEXR` or `OIIO` or `NONE` for the apps image backend?

File: src/apps/ocioconvert/main.cpp
Comment (remia) [reply src/apps/ocioconvert/main.cpp]: Done.

File: docs/guides/contributing/repository_structure.rst
Comment (remia) [reply docs/guides/contributing/repository_structure.rst]: Done.

File: docs/quick_start/for_devs.rst
Comment (remia) [reply docs/quick_start/for_devs.rst:11]: Agreed, I restored the note.

**Comment by remia** [comment]
Here are some notes on the current behaviour of the ImageIO class, feedback welcome:
* Process and write back only the first R,G,B,[A] channels,
* OpenEXR display/data window are not altered during pixel processing,
* There is a slight difference in behaviour between OIIO and EXR backend. OIIO will preserve the input bitdepth when writing the file back and EXR will use the processing bitdepth (eg. FLOAT for `ocioconvert --gpu`),
* `ocioconvert` `--croptofull` and `--ch` only supported for OIIO backend. My tests on `main` seem show that the former is broken on some images and the latter have no effect. We may consider removing this functionality and pointing users to `oiiotool`.

The goal was to keep everything simple and not write a comprehensive image conversion tool.

File: share/cmake/modules/FindOpenEXR.cmake
Comment (remia) [reply share/cmake/modules/FindOpenEXR.cmake:29]: @doug-walker Maybe I misinterpreted your comment and you were more concerned with the `and statically-linked into libOpenColorIO at build time` part, which I should remove?

Comment (doug-walker) [reply share/cmake/modules/FindOpenEXR.cmake:29]: @remia , yes, I was actually thinking about the linking part.  But your question about CMake is a good one too.  These would be great questions to ask during the TSC meeting on Monday.

**Comment by doug-walker** [comment]
@remia , we should ask at the TSC, but I agree it is reasonable to remove the croptofull/ch functionality and point users to oiiotool for that.  

Regarding the output bit-depth, I think your ImageIO class has the necessary functionality, to request a bit-depth.  But we will need to integrate this PR with PR #1648, since in general the output bit-depth may need to be different from the input bit-depth.  (Minor point, but I don't think it is strictly accurate that OIIO will preserve the input depth since the user could have an EXR coming in and a JPG going out.  But I don't think it matters for the purposes of this PR.)

File: share/cmake/modules/FindOpenEXR.cmake
Comment (doug-walker) [reply share/cmake/modules/FindOpenEXR.cmake:29]: I think it's a nice feature to auto download EXR, like some of the other dependencies.  But I don't see why we would ever want to link it into the OCIO library since OCIO core does not depend on it, only the apps.

Comment (remia) [reply share/cmake/modules/FindOpenEXR.cmake]: @doug-walker I updated the comment as it was indeed incorrect and probably a left over from copy & paste.

**Comment by remia** [comment]
Thanks @doug-walker, I'll add an item to the agenda and look into adding a user parameter to control the output bitdepth on write.

File: share/cmake/modules/FindExtPackages.cmake
Comment (lgritz) [inline share/cmake/modules/FindExtPackages.cmake:113]: Same comment here -- any 3.1.x should do (maybe even any 3.x?), I don't think there's a good reason to insist on 3.1.5 and no earlier.

File: share/cmake/modules/FindExtPackages.cmake
Comment (lgritz) [inline share/cmake/modules/FindExtPackages.cmake:121]: So the logic here is that if OIIO is found, it will use it.

I wonder if a better behavior would be to default to using OpenEXR (even in the presence of OIIO), but having a build option that allows people to select OIIO.

Remember, a longstanding goal is to remove the circular dependency. I worry that for the common case of people having both packages on their system, this logic will preserve the circular dependency and that might come back to bite them.

Imagine the following scenario: An older OIIO and OCIO are already on the system. They desire to build a new OCIO, followed by a new OIIO and uses the new OCIO. But here what will happen is that they'll end up with an OCIO that builds against the old OIIO they no longer want to use. Is that problematic? There's no easy way, short of un-installing OIIO ahead of building OCIO, to ensure that the OCIO build doesn't use the OIIO that they don't want.

My recommendation would be to make the circular dependency entirely "opt-in." Or at the very least, make an option that allows them to easily opt out of the circular dependency, even if OIIO can be found.

Comment (KelSolaar) [reply share/cmake/modules/FindExtPackages.cmake:121]: > My recommendation would be to make the circular dependency entirely "opt-in." Or at the very least, make an option that allows them to easily opt out of the circular dependency, even if OIIO can be found

Yes please!

Comment (remia) [reply share/cmake/modules/FindExtPackages.cmake]: Thanks for the feedback, I agree with all being said. There is now a CMake option called `OCIO_USE_OIIO_FOR_APPS` to explicitly opt-in to use OIIO, otherwise it will default on OpenEXR.

File: share/cmake/modules/FindExtPackages.cmake
Comment (remia) [reply share/cmake/modules/FindExtPackages.cmake:41]: You are right, this is a minimum, I just tested with a system 3.1.4 Imath and it is rejecting it as unsuitable. I agree that we should only specify the 3.1 version as listed in the VFX reference platform. The issue I need to sort out is that this version number is also used to dictate what OCIO will build itself when requested (depending on the flag `OCIO_INSTALL_EXT_PACKAGES`).

File: share/cmake/modules/FindExtPackages.cmake
Comment (remia) [inline share/cmake/modules/FindExtPackages.cmake:99]: In the spirit of further cleanup for OCIO 2.2.x, would that be a good time to scrap our custom FindOpenImageIO.cmake script and potentially also increase the minimum OIIO version? OIIO will be used for OSL unit tests and for apps if explicitly asked for. Any thoughts @lgritz?

Comment (remia) [reply share/cmake/modules/FindExtPackages.cmake]: This seems to generate further conflict between our in build Imath and system Imath OIIO is looking for. This will need further investigation. Note that even with the current Find CMake script, the change to OpenEXR by default also introduced some [break](https://github.com/AcademySoftwareFoundation/OpenColorIO/runs/6737918794?check_suite_focus=true#step:8:930) with conflict between OIIO / OSL / OpenEXR.

File: share/cmake/modules/FindExtPackages.cmake
Comment (remia) [reply share/cmake/modules/FindExtPackages.cmake]: I now added a dedicated variable to track the version ExternalProject will build and reverted the find_package calls to 3.1 for both Imath and OpenEXR.

File: share/cmake/modules/FindExtPackages.cmake
Comment (remia) [reply share/cmake/modules/FindExtPackages.cmake]: Is there a reason we would go below 3.1 as the VFX reference platform require this version for 2022 which is our target for the next release?

Comment (lgritz) [reply share/cmake/modules/FindExtPackages.cmake]: In this case, probably not. My impression is that anybody not on 3.1 is probably still stuck on 2.4. I don't think 3.0 was ever important to anybody, since it fell between vfx platform years.

On the other hand, there is almost certainly no difference between 3.0 and 3.1 that is relevant to OCIO, so insisting on 3.1+ rather than 3.0+ has no upside at all for OCIO, while potentially constraining somebody downstream (though probably not, in practice).

Comment (remia) [reply share/cmake/modules/FindExtPackages.cmake]: Thanks for the feedback @lgritz, unless asked otherwise, I'll downgrade to 3.0 is the next round of commits on the PR.

File: share/cmake/modules/FindExtPackages.cmake
Comment (remia) [reply share/cmake/modules/FindExtPackages.cmake:98]: I'm reverting the changes and will try to resolve the Half header conflict mentioned above.

**Comment by remia** [comment]
I think the PR is now ready again for reviews, apologies for the force push 2 weeks ago, I realised later it will probably have broke the reviewers progress on GH.

In terms of changes, the output bitdepth handling on the CPU path of ocioconvert and the build breaks with OpenImageIO have been hopefully fixed. From now on, OCIO will only accept an OIIO version that has been compiled with OpenEXR / Imath 3, I also had to update the minimal requested version to 2.2.14 to be able to find this information. If this is accepted, we will also need @michdolan to update the required checks list on GH due to a naming change (include whether OIIO is used for apps or not).

The analysis builds are currently passing but not fully used due to some missing tags on Imath release branch, limited to 3.1.3 whereas OpenEXR goes to 3.1.5 which create conflicts. I'm trying to get an Imath maintainer to fix that at the moment.

**Comment by zachlewis** [comment]
I just wanted to say, I had an opportunity to refactor our facility "OpenColorIO_tools" package to test this PR, and it feels _so refreshing_ to be able to statically build all the command line tools without the OIIO business. And the OpenEXR build goes as smoothly as the rest of our fancy buildy stuff.

Thanks again Remi -- this makes maintenance so much easier.

**Review by michdolan — APPROVED** [review]
Awesome work @remia !

**Comment by michdolan** [comment]
I can update the required checks after this merges.

**Comment by remia** [comment]
Thanks for the review @michdolan! Could you merge the PR when you have time and update the required checks name at the same time?

FYI there seem to be errors in the analysis builds, one on macOS which seem to track down to using LLVM14 with the current OSL (https://github.com/AcademySoftwareFoundation/OpenShadingLanguage/pull/1492) and the other being a doc building error on Linux which I don't really understand at the moment. Can be looked at at a later time anyway.

**Pull request merged**
